### PR TITLE
Dataflow: Improve performance in flow-through pruning

### PR DESCRIPTION
**Pull request by aschackmull**
Restrict the virtual dispatch relation as much as possible when joining with returning flow that is about to be summarized as flow-through.

### CODE REVIEW

**Comment by aschackmull** [comment]
https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/1405/
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1074/
https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2023/
https://jenkins.internal.semmle.com/job/Changes/job/Python-Differences/544/

**Comment by aschackmull** [comment]
017bf68906ad9418b087ff6c5f40752d8e1140dc:
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1092/
https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2039/

**Comment by aschackmull** [comment]
Looks like there's a possible join-on-configuration problem in `flowThroughOutOfCall`. That will need a fix.

**Comment by aschackmull** [comment]
New round with latest commit:
https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/1422/
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1104/
https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2050/
https://jenkins.internal.semmle.com/job/Changes/job/Python-Differences/561/

**Comment by jbj** [comment]
[CPP-Differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/2050/) looks good. There's a major speedup in `SignedOverflowCheck.ql` that ought to have nothing to do with data flow, so that's probably noise. The speedups in `TaintedPath.ql`, `UncontrolledFormatStringThroughGlobalVar.ql` and `NewArrayDeleteMismatch.ql` could very likely come from data-flow improvements.

**Comment by hvitved** [comment]
Replacement [CSharp-Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/1106/): Performance appears to be unchanged.

**Comment by aschackmull** [comment]
Looks like generally unchanged perf in most projects (which isn't too surprising, as this shores up a weakness in a corner-case of field flow).

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll [hunk 1]
```diff
@@ -1006,6 +1027,27 @@ private module Stage2 {
...
+        pragma[only_bind_into](config)) and
+      fwdFlowOutFromArg(call, node, argAp0, ap, config) and
+      fwdFlowIsEntered(call, pragma[only_bind_into](cc), pragma[only_bind_into](argAp), argAp0,
+        pragma[only_bind_into](config))
+    )
+  }
+
+  pragma[nomagic]
+  private predicate flowThroughIntoCall(
+    DataFlowCall call, ArgNode node1, ParamNode node2, boolean allowsFieldFlow, Configuration config
```
Comment (hvitved) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll:1043]: `node1` -> `arg`, `node2` -> `p` (or `param`)

**Review by hvitved — COMMENTED** [review]
LGTM, some stylistic comments.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll [hunk 2]
```diff
@@ -605,6 +605,15 @@ private module Stage1 {
     )
   }
 
+  pragma[nomagic]
+  predicate callMayFlowThroughRev(DataFlowCall call, Configuration config) {
+    exists(Node node, boolean toReturn |
+      revFlow(node, toReturn, config) and
+      revFlowInToReturn(call, node, config) and
+      revFlowIsReturned(call, toReturn, config)
```
[the remark this answers is not in the document]
Comment (aschackmull) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll:613]: I find `revFlowIsReturnedFromArg` counter-intuitive: The reverse flow is passing through a return edge, hence `revFlowIsReturned`, but we don't know yet whether that reverse flow is going to reach a parameter nor an argument. We also don't match this up with any forward flow from an argument in this particular `call` - we only look at forward flow inside the callee (and whether that came from a parameter).
The symmetric predicate is btw. called `fwdFlowIsEntered`, which I also find quite precise, so any rename should be reflected there with a symmetric name (and `fwdFlowIsEnteredFromReturn` is quite confusing/wrong).

**Comment by aschackmull** [comment]
Changes applied.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (hvitved) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll:613]: Perhaps we simply need to copy the QL doc from the other `revFlowIsEntered` predicates?
```ql
  /**
   * Holds if an output from `call` is reached in the flow covered by `revFlow`
   * and data might flow through the target callable resulting in reverse flow
   * reaching an argument of `call`.
   */
```

[the remark this answers is not in the document]
Comment (aschackmull) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll:613]: That is the current qldoc. (Except for the stage 1 predicates, which are lagging a bit behind in terms of unification with stages 2, 3, and 4)

[the remark this answers is not in the document]
Comment (aschackmull) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll:613]: I guess it makes sense to fix the stage 1 qldoc to match.

**Review by hvitved — APPROVED** [review]

**Pull request merged**